File: problems/11_graphs/0133_clone_graph/solution.dart
// Problem: 0133. Clone Graph
// Difficulty: Medium

class Node {
  int val;
  List<Node> neighbors;
  Node(this.val, [List<Node>? n]) : neighbors = n ?? [];
}

Node? cloneGraph(Node? node) {
  // TODO: implement
  return null;
}

Node? _build(List<List<int>> adj) {
  if (adj.isEmpty) return null;
  final nodes = List.generate(adj.length, (i) => Node(i + 1));
  for (var i = 0; i < adj.length; i++) {
    nodes[i].neighbors = adj[i].map((j) => nodes[j - 1]).toList();
  }
  return nodes[0];
}

List<List<int>> _adj(Node? root) {
  if (root == null) return [];
  final seen = <int, Node>{};
  final st = <Node>[root];
  while (st.isNotEmpty) {
    final n = st.removeLast();
    if (seen.containsKey(n.val)) continue;
    seen[n.val] = n;
    for (final nb in n.neighbors) st.add(nb);
  }
  final keys = seen.keys.toList()..sort();
  return keys.map((v) {
    final nbs = seen[v]!.neighbors.map((n) => n.val).toList()..sort();
    return nbs;
  }).toList();
}


void main() {
  // Test 1
  // final result = solve(...);
  print('All tests passed!');
}

File: problems/11_graphs/0133_clone_graph/solution.py
"""
Problem: 0133. Clone Graph
Difficulty: Medium
URL: https://leetcode.com/problems/clone-graph/

解法：DFS + Hash 對照表。
"""

from __future__ import annotations

from typing import Optional, Dict, List

class Node:
    def __init__(self, val: int = 0, neighbors: Optional[List["Node"]] = None):
        pass


class Solution:
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        pass


def _build(adj: List[List[int]]) -> Optional[Node]:
    if not adj:
        return None
    nodes = [Node(i + 1) for i in range(len(adj))]
    for i, nbs in enumerate(adj):
        nodes[i].neighbors = [nodes[j - 1] for j in nbs]
    return nodes[0]

def _adj(root: Optional[Node]) -> List[List[int]]:
    if root is None:
        return []
    seen: Dict[int, Node] = {}
    stack = [root]
    while stack:
        n = stack.pop()
        if n.val in seen:
            continue
        seen[n.val] = n
        for nb in n.neighbors:
            stack.append(nb)
    out = []
    for v in sorted(seen):
        out.append(sorted(nb.val for nb in seen[v].neighbors))
    return out

if __name__ == "__main__":
    node = Node()

    # Test 1
    # result = node.solve(...)
    # assert result == expected

    print("All tests passed!")
